### PR TITLE
correction of `IsMonomial`, mainly for reducible characters

### PR DESCRIPTION
**Pull request by ThomasBreuer**
correction of `IsMonomial`, mainly for reducible characters

- change the `TestMonomialQuick` and `TestMonomial` methods
  for characters such that also reducible characters are handled correctly

- restrict the `IsPrimitiveCharacter` method that relies on
  quasi-primitivity to irreducible characters

- define `IsSubnormallyMonomial` only for irreducible characters

- add tests

### CODE REVIEW

File: lib/ctblmono.gi [hunk 1]
```diff
@@ -357,6 +364,10 @@ end );
 ##
 #M  TestQuasiPrimitive( <chi> ) . . . . . . . . . . . . . . . for a character
 ##
+##  This works also for reducible <chi>.
+##  Note that a representation affording <chi> maps the centre of <chi>
+##  to scalar matrices.
```
Comment (fingolfin) [inline lib/ctblmono.gi]: What is the relevance of this comment (on the centre) at this point? To me, a comment here is for somebody who wants to call the function without caring what it does... Wouldn't this factoid about the centre be more helpful below, perhaps near the "proof" comment?

File: lib/ctblmono.gi [hunk 2]
```diff
@@ -1013,6 +1015,8 @@ InstallMethod( IsMonomialNumber,
 ##
 #M  TestMonomialQuick( <chi> )  . . . . . . . . . . . . . . . for a character
 ##
+##  We may assume that <chi> is an irreducible character.
```
Comment (fingolfin) [inline lib/ctblmono.gi]: "We may assume"? This makes it sound as if it was optional, or as if there was some underlying reason... Remove "may"?, resp. say "Assumes that..." ?

Or even clarify: "This function assumes that <chi> is an irreducible character, but does not verify this. Other inputs may thus leading to nonsensical output." ?

File: lib/ctblmono.gi [hunk 3]
```diff
@@ -1330,38 +1344,38 @@ InstallMethod( TestMonomial,
 
     Info( InfoMonomial, 1, "TestMonomial called" );
 
-    # Start wirth elementary tests for monomiality.
-    test:= TestMonomialQuick( chi );
+    # Start with elementary tests for monomiality.
+    if   IsIrreducibleCharacter( chi ) then
+      test:= TestMonomialQuick( chi );
+    elif HasIsMonomialCharacter( chi ) then
```
Comment (fingolfin) [inline lib/ctblmono.gi]: Should that `HasIsMonomialCharacter` perhaps be first, given that it'll be the quickest test by far?

File: lib/ctblmono.gi (hunk 1 above)
Comment (ThomasBreuer) [reply lib/ctblmono.gi]: No.
Somebody who wants to call the function should look at the documentation
in the corresponding `.gd` file.
Comments in the implementation part are intended for somebody who wants to look at the code.

File: lib/ctblmono.gi (hunk 2 above)
Comment (ThomasBreuer) [reply lib/ctblmono.gi]: O.k.

File: lib/ctblmono.gi (hunk 3 above)
Comment (ThomasBreuer) [reply lib/ctblmono.gi]: O.k.

**Comment by fingolfin** [comment]
Something went wrong there -- you now have two commits with identical (?) content and a merge commit in this PR. Perhaps try to `git rebase master` (if that fails with an error, you can always `git rebase --abort`)?

**Comment by fingolfin** [comment]
@ThomasBreuer Unfortunately, this PR causes multiple test failures, already in `testinstall`, e.g. `tst/testinstall/ctblmono.tst` now fails. You can reproduce this locally by reading the file `tst/testinstall.g` from the GAP root directory into GAP.

**Comment by fingolfin** [comment]
The tests pass now, but this PR contains duplicates of all kinds of commits on master. You'll have to `git rebase master` once again -- and make sure your `master` really is in sync with the `master` branch of `gap-system/gap`...

**Pull request merged**